pkg/bff/server.go

```diff
@@ -221,38 +217,59 @@ func (s *Server) SetURL(url string) {
 	log.Debug().Str("url", url).Msg("server url set")
 }
 
-// Middleware that tracks request performance with Sentry.
-func (s *Server) TrackPerformance() gin.HandlerFunc {
-	return func(c *gin.Context) {
-		span := sentry.StartSpan(c.Request.Context(), c.Request.URL.Path)
-		defer span.Finish()
-		c.Next()
+func (s *Server) setupRoutes() (err error) {
+	// Instantiate authentication middleware
+	var authenticator gin.HandlerFunc
+	if authenticator, err = auth.Authenticate(s.conf.Auth0); err != nil {
+		return err
 	}
-}
 
-func (s *Server) setupRoutes() (err error) {
-	// Application Middleware
-	s.router.Use(ginzerolog.Logger("gin"))
-
-	// Gin middleware needs to be added before Sentry for correct recovery handling
-	s.router.Use(gin.Recovery())
-	s.router.Use(sentrygin.New(sentrygin.Options{
-		Repanic:         true,
-		WaitForDelivery: false,
-	}))
+	var tracing gin.HandlerFunc
 	if s.conf.Sentry.TrackPerformance {
-		s.router.Use(s.TrackPerformance())
+		tracing = s.TrackPerformance()
+	}
+
+	// Application Middleware
+	// NOTE: ordering is very important to how middleware is handled.
+	middlewares := []gin.HandlerFunc{
+		// Logging should be outside so we can record the complete latency of requests.
+		// NOTE: logging panics will not recover.
+		logger.GinLogger("bff"),
+
+		// Panic recovery middleware; note: gin middleware needs to be added before sentry
+		gin.Recovery(),
+		sentrygin.New(sentrygin.Options{
+			Repanic:         true,
+			WaitForDelivery: false,
+		}),
+
+		// Tracing helps us with our peformance metrics and should be as early in the
+		// chain as possible. It is after recovery to ensure trace panics recover.
+		tracing,
+
+		// CORS configuration allows the front-end to make cross-origin requests.
+		cors.New(cors.Config{
+			AllowOrigins:     s.conf.AllowOrigins,
+			AllowMethods:     []string{"GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"},
+			AllowHeaders:     []string{"Origin", "Content-Length", "Content-Type", "Authorization", "X-CSRF-TOKEN", "sentry-trace"},
+			AllowCredentials: true,
+			MaxAge:           12 * time.Hour,
+		}),
+
+		// Maintenance mode handling - does not require authentication.
+		s.Available(),
+
+		// Authentication happens as late as possible; all middleware after this should
+		// require a user context; if it doesn't, it should come before authentication.
+		authenticator,
+	}
+
+	// Add the middleware to the router
+	for _, middleware := range middlewares {
+		if middleware != nil {
+			s.router.Use(middleware)
+		}
 	}
```

Comment:
Above code reorganizes how middleware is added to give us a bit more control of ordering and visibility into the entire middleware chain.